filebeat/input/filestream/input.go

```diff
@@ -167,7 +180,53 @@ func (inp *filestream) Run(
 
 	// The caller of Run already reports the error and filters out errors that
 	// must not be reported, like 'context cancelled'.
-	return inp.readFromSource(ctx, log, r, fs.newPath, state, publisher, metrics)
+	err = inp.readFromSource(ctx, log, r, fs.newPath, state, publisher, metrics)
+	if err != nil {
+		deleteFile := false
+		switch {
+		case errors.Is(err, io.EOF):
+			if inp.deleterConfig.OnClose.EOF {
+				log.Infof("'%s' will be removed because 'delete.on_close.eof' is set", fs.newPath)
+				deleteFile = true
+			}
+		case errors.Is(err, ErrInactive):
+			if inp.deleterConfig.OnClose.Inactive {
+				log.Infof("'%s' will be removed because 'delete.on_close.inactive' is set", fs.newPath)
+				deleteFile = true
+			}
+		}
+
+		if deleteFile {
+			// Use a goroutine not to lock the harvester shutdown process
+			go func(path string) {
+				for !cursor.Finished() {
+					log.Debugf("'%s' is not finished, waiting...", fs.newPath)
+					time.Sleep(inp.deleterConfig.RetryTick)
+				}
+				if err := os.Remove(path); err != nil {
+					log.Errorf("could not remove '%s', retrying in 2s. Error: %s", path, err)
+					maxWait := inp.deleterConfig.RetryTimeout
+					ticker := time.NewTicker(inp.deleterConfig.RetryTick)
+					defer ticker.Stop()
+					for maxWait >= 0 {
+						maxWait -= inp.deleterConfig.RetryTick
+						select {
+						case <-ticker.C:
```

Comment:
I believe the case with `ctx.Cancelation.Done()` belongs here 🙂

Reply:
I'll add it, thanks!